src/main/java/spiffe/provider/CertificateUtils.java

```diff
@@ -0,0 +1,176 @@
+
+    /**
+     * Generate the collection of X509Certificates
+     *
+     * @param input as byte array
+     * @return a Set of X509Certificate
+     * @throws CertificateException
+     */
+    static Set<X509Certificate> generateCertificates(byte[] input) throws CertificateException {
+        Collection<? extends Certificate> certificates =  getCertificateFactory().generateCertificates(new ByteArrayInputStream(input));
+        return certificates.stream().map(c -> (X509Certificate) c).collect(toSet());
+    }
+
+    /**
+     * Generate a single X509Certificate
+     *
+     * @param input as byte array
+     * @return an instance of X509Certificate
+     * @throws CertificateException
+     */
+    static X509Certificate generateCertificate(byte[] input) throws CertificateException {
+        return (X509Certificate) getCertificateFactory().generateCertificate(new ByteArrayInputStream(input));
+    }
+
+    /**
+     * Generates a PrivateKey from the X509SVIDKey ByteArray
+     *
+     * It uses PKCS8EncodedKeySpec that represents the ASN.1 encoding of a private key
+     *
+     * @param input as byte array
+     * @return
+     * @throws NoSuchAlgorithmException
+     * @throws InvalidKeySpecException
+     */
+    static PrivateKey generatePrivateKey(byte[] input) throws NoSuchAlgorithmException, InvalidKeySpecException {
+        KeyFactory keyFactory = KeyFactory.getInstance("EC", new BouncyCastleProvider());
+        return keyFactory.generatePrivate(new PKCS8EncodedKeySpec(input));
+    }
+
+    /**
+     * Validate a certificate chain against a set of trusted certificates.
+     *
+     * @param chain    certificate chain
+     * @param trustedCerts
+     * @throws CertificateException
+     */
+    static void validate(X509Certificate[] chain, Set<X509Certificate> trustedCerts) throws CertificateException {
+        PKIXParameters pkixParameters = toPkixParameters(trustedCerts);
+        CertPath certPath = certificateFactory.generateCertPath(asList(chain));
+        try {
+            certPathValidator.validate(certPath, pkixParameters);
+        } catch (CertPathValidatorException | InvalidAlgorithmParameterException e) {
+            throw new CertificateException(e);
+        }
+    }
+
+    /**
+     * Validates that the SPIFFE ID is present and matches the SPIFFE ID configured in
+     * the java.security property ssl.spiffe.accept
+     *
+     * @param chain an array of X509Certificate that contains the Peer's SVID to be validated
+     * @throws CertificateException when either the certificates doesn't have a SPIFFE ID or the SPIFFE ID is not trusted
+     */
+    static void checkSpiffeId(X509Certificate[] chain) throws CertificateException {
+        Optional<String> spiffeId = getSpiffeId(chain[0]);
+        if (spiffeId.isPresent()) {
+            String acceptedSpiffeId = Security.getProperty("ssl.spiffe.accept");
+
+            if (!StringUtils.equals(spiffeId.get(), acceptedSpiffeId)) {
```

Comment:
If spiffeID from the certificate is blank it will not pass the condition _isPresent_, and if the _acceptedSpiffeId_ is blank it means that no spiffe ID will be trusted and it will always throw the CertificateException.

Reply:
actually IsPresent implementation takes care only of null:
```
 public boolean isPresent() {
        return value != null;
    }
```
no empty string,